### DISCUSSION TITLE
Admin REST API v2

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#8983] [Ideas]
The current Admin REST APIs are not very user friendly, and poorly documented. There are a number of issues with the APIs that require breaking changes, such as:

* Pagination requires separate call to retrieve count
* Generated IDs for created resources are returned in Location header, which can be hard to deal with
* Mixing create and update in post method
* Inconsistencies throughout the APIs
* Lack of versioning making it impossible to handle breaking changes to the APIs
* Insufficient validation

The REST API documentation is also lacking in quite a lot of aspects, and there's also lack of an OpenAPI specification.

We really should complete the REST API guidelines we started on, to make sure we have consistency throughout all REST APIs we offer (excluding of course OIDC and SAML, and other spec based APIs). This should be covered in https://github.com/keycloak/keycloak/discussions/8499

Then we should break the Admin API into multiple smaller APIs such as:

* Realm API
* Client API
* User API

These smaller APIs can be implemented and versioned separately.

**Comment by edewit** [comment]
Let's start by creating a OpenAPI specification. We could generate clients and server code based on this. Maybe we could start with using something like https://studio.apicur.io/ to create a design?

**Comment by patrykbaszak** [comment]
Is something going on about this?

**Comment by anowak-ct** [comment]
A description about how Paging mechanism is completely missing in the admin REST API docs. E.g. The number of results for getting list of users is limited by default top 100. The doc is mentioning `first` and `max` url parameters for paging, but there's no doc about how to go thru consecutive pages. A section about paging in the doc would be quite helpful.

**Reply by jonkoops** [reply]
Yes, this is a big problem we need to address in v2. The current API doesn't even implement paging the same way consistently across endpoints.
